### PR TITLE
[LIVY-972] Add GitHub Action for CI image

### PR DESCRIPTION
**Pull request by dacort**
## What changes were proposed in this pull request?

- Adds Several GitHub Workflows
  - [x] CI image builder
  - [x] Unit tests for `git push`
  - [x] Integration tests for PRs

## How was this patch tested?

Manual validation in the course of the pull request and new GitHub Actions workflows.

### CODE REVIEW

File: .github/workflows/unit-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  build:
+    runs-on: ubuntu-20.04
+    container: ghcr.io/${{ github.repository_owner }}/livy-ci:latest
+    steps:
+    - 
+      name: Checkout
+      uses: actions/checkout@v3
+    - 
+      name: Build with Maven
+      run: mvn -Pthriftserver -DskipITs -Dmaven.javadoc.skip=true -B -V -e verify
```
Comment (ksumit) [inline .github/workflows/unit-tests.yaml:29]: I'm assuming we will add more jobs that cover other scenarios?

**Review by ksumit — COMMENTED** [review]
Thanks for taking this up.

File: .github/workflows/build-ci-image.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+name: 'Build CI images'
+on: 
+  push:
+    branches: ["main", "feature/github-actions"]
```
[the remark this answers is not in the document]
Comment (dacort) [reply .github/workflows/build-ci-image.yaml:20]: Yes, I'll add a `path` on this that limits the action to if the `./dev/docker/Dockerfile` changes and also remove the feature branch - had that here so I could test the action.

File: .github/workflows/unit-tests.yaml (hunk 1 above)
Comment (dacort) [reply .github/workflows/unit-tests.yaml:29]: Yes, planning on adding the other pre-existing jobs from `.travis.yml` file, which includes thriftserver unit tests and Spark 3.x tests for each.

File: .github/workflows/integration-tests.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      name: Checkout
+      uses: actions/checkout@v3
+    - 
+      name: Cache local Maven repository
+      uses: actions/cache@v3
+      with:
+        path: |
+          /root/.m2/repository
+          !/root/.m2/repository/org/apache/livy
+        key: ${{ runner.os }}-maven-${{ hashFiles('pom.xml', '*/pom.xml', 'thriftserver/*/pom.xml', 'core/*/pom.xml', 'repl/*/pom.xml', 'scala-api/*/pom.xml') }}
```
Comment (ksumit) [inline .github/workflows/integration-tests.yaml:42]: does `hashFiles` support `**/pom.xml`?

**Review by ksumit — COMMENTED** [review]
Would it make sense to merge `integration-tests.yaml` and `unit-tests.yaml` and use some sort of flag to use one vs the other. We could also avoid building the repository two times (though I don't think that's a big problem since the builds are happening in parallel and probably not adding to the runtime of the overall pipelines). Thoughts?

File: .github/workflows/integration-tests.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+env:
+  MAVEN_OPTS: -Dhttp.keepAlive=false -Dmaven.wagon.http.pool=false -Dmaven.wagon.http.retryHandler.class=standard -Dmaven.wagon.http.retryHandler.count=3
+jobs:
+  build:
+    runs-on: ubuntu-20.04
+    # TODO: Possibly point to the ./build-ci-image.yaml with the "uses" key
+    container: ghcr.io/${{ github.repository_owner }}/livy-ci:latest
+    strategy:
+      matrix:
+        spark_version: ["2.4", "3.0"]
```
Comment (ksumit) [inline .github/workflows/integration-tests.yaml:30]: I think we can add the unit-test vs integration-test logic here.

File: .github/workflows/integration-tests.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+name: Integration Tests
+on:
+  pull_request:
+    types: [opened, reopened, synchronize]
+env:
+  MAVEN_OPTS: -Dhttp.keepAlive=false -Dmaven.wagon.http.pool=false -Dmaven.wagon.http.retryHandler.class=standard -Dmaven.wagon.http.retryHandler.count=3
```
[the remark this answers is not in the document]
Comment (dacort) [reply .github/workflows/integration-tests.yaml:22]: Unfortunately, yes. If the Docker build is busy for a while and doesn't download anything, the http pool times out apparently. There have been other projects that resolved this in a similar way like [pulsar](https://github.com/apache/pulsar/pull/8386).

File: .github/workflows/unit-tests.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+name: Unit Tests
+on: [push]
+env:
+  MAVEN_OPTS: -Dhttp.keepAlive=false -Dmaven.wagon.http.pool=false -Dmaven.wagon.http.retryHandler.class=standard -Dmaven.wagon.http.retryHandler.count=3
```
[the remark this answers is not in the document]
Comment (dacort) [reply .github/workflows/unit-tests.yaml:20]: Yep.

File: .github/workflows/build-ci-image.yaml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+name: 'Build CI images'
+on: 
+  push:
+    branches: ["main"]
+    paths:
+    - 'dev/docker/Dockerfile'
```
[the remark this answers is not in the document]
Comment (dacort) [reply .github/workflows/build-ci-image.yaml:22]: 1. Yes, good idea, will add that.
2. Hm, good point. I had the `main` branch here so we wouldn't push the image unless it got merged into main, but would be good to be able to validate the image as well before doing so.

**Comment by dacort** [comment]
> Would it make sense to merge `integration-tests.yaml` and `unit-tests.yaml`

I wasn't quite sure if there was an easy way to do that. Can definitely look more into it, but ideas welcome.

**Comment by ksumit** [comment]
> > Would it make sense to merge `integration-tests.yaml` and `unit-tests.yaml`
> 
> I wasn't quite sure if there was an easy way to do that. Can definitely look more into it, but ideas welcome.

Here is an example that i found on https://docs.github.com/en/actions/learn-github-actions/contexts
```
name: CI
on: push
jobs:
  prod-check:
    if: ${{ github.ref == 'refs/heads/main' }}
    runs-on: ubuntu-latest
    steps:
      - run: echo "Deploying to production server on branch $GITHUB_REF"
```

File: .github/workflows/integration-tests.yaml (hunk 4 above)
Comment (ksumit) [reply .github/workflows/integration-tests.yaml:30]: @dacort I wouldn't want to hold this PR for minor things. If you feel comfortable with the current state, I would push it as it is (something better than nothing) and then iterate on the improvements later on. Let me know what you think.

**Comment by gyogal** [comment]
Thanks a lot for working on this @dacort! I agree with @ksumit's points and if you feel like this is good to go as is, we could merge it to restore CI functionality for the project.
Post-merge would it be possible to run this for PRs received recently where the builds were not kicked off (such as #396)?

**Comment by dacort** [comment]
Thanks @gyogal - Yea, @ksumit I think for now we should merge this. I imagine the PR workflow and unit-test workflow will diverge in the future so would rather get this merged and iterate.

@gyogal re: run the checks retroactively, I don't think it'll happen automatically but if folks push another commit it should.

**Review by lmccay — APPROVED** [review]
+1

**Review by gyogal — APPROVED** [review]

**Pull request merged**